src/npm-resolve.ts

```diff
@@ -1,13 +1,124 @@
+import { createDeno } from './deno.ts';
 import { PluginConfig } from './types.ts';
 
-export default function npmResolve({}: PluginConfig) {
+export default function npmResolve(config: PluginConfig) {
+  // TODO(bartlomieju): both plugins should use the same instance, or maybe
+  // both plugins should be combined.
+  const deno = createDeno(config);
+
   return {
     name: 'vite:deno-npm-resolve',
 
     enforce: 'pre' as const,
 
-    resolveId() {
+    async resolveId(importee: string, importer: string | undefined) {
+      console.log('resolveId npm', importee, importer);
+
+      if (importee.startsWith('npm:')) {
+        await deno.cache(importee);
+        return importee;
+      }
+
       return null;
     },
+
+    async load(id: string) {
+      console.log('load npm', id);
+
+      if (!id.startsWith('npm:')) {
+        return null;
+      }
+
+      const ref = npmPackageReference(id);
+
+      // TODO(bartlomieju): this shouldn't be called on each load, it can
+      // be cached per run
+      const cacheInfo = await deno.cacheInfo();
+      const npmCache = cacheInfo.npmCache;
+
+      const specifier = ref.name;
+      const version = ref.versionReq;
+
+      if (!version) {
+        throw new Error(`Version not specified for ${specifier}`);
+      }
+
+      console.log('NpmPackageReference', ref);
+      const moduleDirPath =
+        `${npmCache}/registry.npmjs.org/${specifier}/${version}`;
+
+      const packageJson = JSON.parse(
+        await Deno.readTextFile(
+          `${moduleDirPath}/package.json`,
+        ),
+      );
+
+      let file = packageJson.main || 'index.js';
+      if (ref.subPath) {
+        // TODO(bartlomieju): handle other extensions
+        file = `${ref.subPath}.js`;
```

Comment:
This is only one facet of multiple ways a package entrypoint can be defined. If Deno does this internally (I assume it would have to?) perhaps we just focus on getting the Deno side fixed? Else it's a lot of rework here.

Reply:
Maybe, but it will be slow - calling a subprocess to resolve each specifier seems worse than porting some code/using a package that can do that.

Reply:
Good point 😄